src/rust/ide/src/controller/upload.rs

```diff
@@ -0,0 +1,608 @@
+        match self.file.data.next_chunk().await {
+            Ok(Some(data)) => {
+                info!(self.logger, "Received chunk of {self.file.name} of size {data.len()}");
+                trace!(self.logger, "Data: {data:?} uploading to {self.remote_path:?}");
+                self.bin_connection.write_file(&self.remote_path,&data).await?; // TODO replace with write bytes.
+                self.checksum.input(&data);
+                self.bytes_uploaded += data.len();
+                Ok(false)
+            },
+            Ok(None) => {
+                if self.bytes_uploaded != self.file.size {
+                    error!(self.logger, "The promised file size ({self.file.size}) and uploaded \
+                        data length ({self.bytes_uploaded}) do not match. Leaving as much data as \
+                        received.");
+                    self.bytes_uploaded = self.file.size;
+                }
+                self.check_checksum().await?;
+                Ok(true)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    async fn check_checksum(&mut self) -> FallibleResult {
+        let remote = self.json_connection.file_checksum(&self.remote_path).await?.checksum;
+        let local  = std::mem::take(&mut self.checksum).into();
+        if remote != local {
+            Err(ChecksumMismatch {remote,local}.into())
+        } else {
+            Ok(())
+        }
+    }
+}
+
+
+
+// ==================================
+// === NodeFromDroppedFileHandler ===
+// ==================================
+
+/// The handler for nodes created by dragging and dropping files into IDE.
+///
+/// It is responsible for creating node, uploading file and updating the node's metadata.
+#[derive(Clone,CloneRef,Debug)]
+pub struct NodeFromDroppedFileHandler {
+    logger   : Logger,
+    project  : model::Project,
+    graph    : controller::Graph
+}
+
+impl NodeFromDroppedFileHandler {
+    /// Constructor
+    pub fn new(parent:impl AnyLogger, project:model::Project, graph:controller::Graph) -> Self {
+        let logger = Logger::sub(parent,"NodeFromDroppedFileHandler");
+        Self{logger,project,graph}
+    }
+
+    /// Create a node from dropped file and start uploading file.
+    ///
+    /// The function returns once the node is created; the uploading process is scheduled in the
+    /// global executor. The node's metadata will be updated with the uploading progress and
+    /// error messages if any.
+    pub fn create_node_and_start_uploading
+    (self, file:FileToUpload<impl DataProvider + 'static>, position:Position) -> FallibleResult {
+        let node = self.graph.add_node(Self::new_node_info(&file,position))?;
+        executor::global::spawn(async move {
+            if let Err(err) = self.upload_file(node,file).await {
+                error!(self.logger, "Error while uploading file: {err}");
+                self.update_metadata(node, |md| md.error = Some(err.to_string()));
```

Comment:
If we do not want this to appear on undo stack, we need to opt-out by creating an aborted transaction.